File: 프로그래머스/lv3/42895. N으로 표현/N으로 표현.py
d=[set() for _ in range(9)]
def go(x,cnt):
    d[cnt].add(x)
    for i in range(1,cnt):
        for j in d[i]:
            for k in d[cnt-i]:
                d[cnt].add(j+k)
                d[cnt].add(j-k)
                d[cnt].add(j*k)
                if k:
                    d[cnt].add(j//k)

def solution(N, number):
    for i in range(1,9):
        a=int(str(N)*i)
        go(a,i)
        if number in d[i]:
            return i
    return -1

File: 프로그래머스/lv3/42895. N으로 표현/N으로 표현.cpp
#include <string>
#include <vector>
using namespace std;
#define MAX 987654321
int res=MAX;
int n;
int num;
void solve(int sum,int cnt)
{
    int i;
    if (res<=cnt||cnt>8)
        return;
    if (sum==num)
        res=min(cnt,res);
    int temp=0;
    for (i=cnt;i<8;i++)
    {
        temp=temp*10+n;
        solve(sum+temp,i+1);
        solve(sum-temp,i+1);
        solve(sum/temp,i+1);
        solve(sum*temp,i+1);
    }
}
int solution(int N, int number) {
    int answer = 0;
    n=N;
    num=number;
    solve(0,0);
    if (res==MAX)
        res=-1;
    return res;
}